invenio_files_rest/models.py

```diff
@@ -1187,6 +1264,118 @@ def relink_all(cls, old_file, new_file):
+            bucket=self.bucket, key=key, desc=True)
+        dest_head = ObjectVersion.query_heads_by_bucket(bucket=bucket).filter(
+            ObjectVersion.key == key).one_or_none()
+        if dest_head and all(orig.file_id != dest_head.file_id
+                             for orig in orig_ancestor):
+            # there are two branch, I can't merge!
+            raise MergeConflict()
+        # delete or copy
+        if self.deleted:
+            self.delete(bucket=bucket, key=self.key)
+        else:
+            self.copy(bucket=bucket, key=self.key)
+
+
+class ObjectVersionTag(db.Model):
+    """Model for storing tags associated to object versions.
+
+    Used for storing extra technical information for an object version.
+    """
+
+    __tablename__ = 'files_objecttags'
+
+    version_id = db.Column(
+        UUIDType,
+        db.ForeignKey(ObjectVersion.version_id, ondelete='CASCADE'),
+        default=uuid.uuid4,
+        primary_key=True, )
+    """Object version id."""
+
+    key = db.Column(db.String(255), primary_key=True)
+    """Tag key."""
+
+    value = db.Column(db.Text, nullable=False)
+    """Tag value."""
+
+    object_version = db.relationship(ObjectVersion, backref='tags')
+    """Relationship to object versions."""
+
+    def copy(self, object_version=None, key=None):
+        """Copy a tag to a given object version.
+
+        :param object_version: The object version instance to copy the tag to.
+            Default: current object version.
+        :param key: Key of destination tag.
+            Default: current tag key.
+        :return: The copied object version tag.
+        """
+        return ObjectVersionTag.create(
+            self.object_version if object_version is None else object_version,
+            key or self.key,
+            self.value
+        )
+
+    @classmethod
+    def get(cls, object_version, key):
+        """Get the tag object."""
+        return cls.query.filter_by(
+            version_id=as_object_version_id(object_version),
+            key=key,
+        ).one_or_none()
+
+    @classmethod
+    def create(cls, object_version, key, value):
+        """Create a new tag for a given object version."""
+        with db.session.begin_nested():
+            obj = cls(version_id=as_object_version_id(object_version),
+                      key=key,
+                      value=value)
+            db.session.add(obj)
+        return obj
+
+    @classmethod
+    def create_or_update(cls, object_version, key, value):
+        """Create or update a new tag for a given object version."""
+        obj = cls.get(object_version, key)
+        if obj:
+            obj.value = value
+        else:
+            cls.create(object_version, key, value)
+
+    @classmethod
+    def get_value(cls, object_version, key):
+        """Get the tag value."""
+        obj = cls.get(object_version, key)
+        return obj.value if obj else None
+
+    @classmethod
+    def delete(cls, object_version, key):
```

Comment:
I would add support to delete all keys as well. Essentially allow either not specifying ``key`` as a kwarg or pass None.